src/hooks/use-authentication/index.ts

```diff
@@ -1,53 +1,80 @@
-import { useEffect, useMemo } from 'react'
+import { useEffect, useMemo, useState } from 'react'
 import { useRouter } from 'next/router'
 import { useSession } from 'next-auth/react'
 import { saveAndLoadAnalytics } from '@hashicorp/react-consent-manager'
 import { preferencesSavedAndLoaded } from '@hashicorp/react-consent-manager/util/cookies'
-import { SessionData, UserData, ValidAuthProviderId } from 'types/auth'
+import { AuthErrors, SessionData, ValidAuthProviderId } from 'types/auth'
 import { UseAuthenticationOptions, UseAuthenticationResult } from './types'
 import { makeSignIn, makeSignOut, signUp } from './helpers'
 
 export const DEFAULT_PROVIDER_ID = ValidAuthProviderId.CloudIdp
 
-/**
- * Hook for consuming user, session, and authentication state. Sources all data
- * from next-auth/react's `useSession` hook.
- *
- * https://next-auth.js.org/getting-started/client#usesession
- */
-const useAuthentication = (
-	options: UseAuthenticationOptions = {}
-): UseAuthenticationResult => {
+function useAuthMethods() {
 	// Get router path for `signIn` and `signOut` `callbackUrl`s
 	const router = useRouter()
 
 	// Set up memoized `signIn` and `signOut` callbacks
-	const signIn = useMemo(
-		() => makeSignIn({ routerPath: router.asPath }),
-		[router.asPath]
-	)
+	const signIn = useMemo(() => makeSignIn(), [])
 	const signOut = useMemo(
 		() => makeSignOut({ routerPath: router.asPath }),
 		[router.asPath]
 	)
 
+	return { signIn, signOut }
+}
+
+/**
+ * Hook for consuming user, session, and authentication state. Sources all data
+ * from next-auth/react's `useSession` hook.
+ *
+ * https://next-auth.js.org/getting-started/client#usesession
+ */
+const useAuthentication = (
+	options: UseAuthenticationOptions = {}
+): UseAuthenticationResult => {
+	const { signIn, signOut } = useAuthMethods()
 	// Get option properties from `options` parameter
+	// `onUnauthenticated` function only calls when `isRequired` is true
 	const { isRequired = false, onUnauthenticated = () => signIn() } = options
 
 	// Pull data and status from next-auth's hook, and pass options
 	const { data, status } = useSession({
 		required: isRequired,
 		onUnauthenticated,
 	})
+	const [isAuthenticated, setIsAuthenticated] = useState(false)
+
+	// retain authenticated state based on status
+	useEffect(() => {
+		if (!isAuthenticated && status === 'authenticated') {
+			setIsAuthenticated(true)
+		} else if (isAuthenticated && status === 'unauthenticated') {
+			setIsAuthenticated(false)
+		} else if (status !== 'authenticated') {
+			setIsAuthenticated(false)
+		}
+	}, [status, isAuthenticated])
```

Comment:
Change here is to store `isAuthenticated` in state so that we can adjust this based not only on the session status but if there is an error, to prevent API calls for authenticated data if we are in an errored state.

Reply:
Open to another way, but wanted to see how this behaved.